research/design-doc/sparse_checkout_asishkumar.md

```diff
@@ -0,0 +1,68 @@
+# KPM sparse checkout
+
+**Author**: Asish Kumar
+
+## Abstract
+
+`kpm` manages third-party libraries through Git repositories, requiring a `kcl.mod` file at the root directory. It treats the entire Git repository as a single `kcl` package, which is inefficient for monorepos containing multiple `kcl` packages. Often, a `kcl` project depends on just one package within a monorepo, but `kpm` downloads the entire repository. Therefore, `kpm` needs to allow adding a subdirectory of a Git repository as a dependency, enabling it to download only the necessary parts and improve performance.
+
+## User Interface
+
+I will add a new flag called `--subdir` in `kcl mod add` command.  This flag will specify the path to the desired subdirectory within the Git repository. Below is the syntax for the enhanced kpm add command:
+
+```
+kcl mod add --git <git-repo-url> --subdir <subdir> 
+```
+
+The `--subdir` flag will be optional. If the flag is not provided, `kpm` will download the entire repository as it does now. If the flag is provided, `kpm` will download only the specified subdirectory of the git repo.
+
+Example usage: 
+
+```
+kcl mod add --git https://github.com/kcl-lang/modules --subdir add-certificates-volume 
+```
+
+This command will download the `add-certificates-volume` subdirectory from the `modules` repository and append it in the subdir array of the `kcl.mod` file.
+
+The `kcl.mod` file will look like this:
+
+```
+[dependencies]
+bbb = { git = "https://github.com/kcl-lang/flask-demo-kcl-manifests.git", commit = "ade147b", subdir = ["something"]}
```

Comment:
Why `subdir` is a list ?

Reply:
I defined this because in the future if we want to add other subdir from a git repo say kcl-lang/modules then we would be able to achieve that without overwritting the previous subdir